### ISSUE TITLE
Switch to TIMESTAMP WITH TIME ZONE for current_* tables

### ISSUE THREAD

**Issue by pnorman** [openstreetmap/openstreetmap-website#375]
Currently the current_nodes, current_ways and current_relations tables include a timestamp column of type `TIMESTAMP WITHOUT TIME ZONE`.

I've asked around and consulted some sources and the advice I got was best summarized as

`19:12 <RhodiumToad> pnorman: unless you're writing a calendaring/scheduling app, timestamp with time zone is pretty much always the way to go`

This is not purely a theoretical concern. Currently there is no way to my knowledge to load to an apidb that does not result in time zone errors. See [this osmosis discussion](http://lists.openstreetmap.org/pipermail/osmosis-dev/2013-July/001586.html)

Neither the rails port nor cgimap returns the correct timestamps right now on any instance other than osm.org running on a server where the local time is not set to UTC, because the problem is in the database.

Cross-reference pnorman/openstreetmap-api-testsuite#1
cc: @brettch, @zerebubuth, @iandees

**Comment by gravitystorm** [comment]
This issue has come up again in #2028 and I feel it's something that we need to resolve. Unfortunately there's a few misunderstandings from @tomhughes and @zerebubuth on this topic, so I hope to persuade them of the case for changing the column type.

http://www.postgresqltutorial.com/postgresql-timestamp/ has some background reading.

The main misunderstanding appears to be that `timestamp without timezone` means UTC and `timestamp with timezone` stores a timestamp with an additional timezone offset. **This is not the case**. Both columns store the same information, i.e. a timestamp. Same bytes. Same timestamp. The difference is only how software should interpret those bytes.

* `timestamp without timezone` - "2018/11/21 10:05" is an event that can happen [38 times](https://www.timeanddate.com/time/current-number-time-zones.html), once in each timezone throughout the day. *"without timezone" means "treat as local time"* and so if you connect to the database with a client in UTC+8, the timestamp refers to "2018-11-21T10:05Z+08:00". This is absolutely not what we want, so everything that connects to our database has to be **very** careful to interpret the timestamp as UTC - and we end up jumping through hoops as described above.
* `timestamp with timezone` - "2018/11/21 10:05" is an event that happens exactly once, at 10:05 UTC. Let me emphasise again that this column type does **not** store a timezone offset. It is an indication that the timestamp is, and always is, UTC. This is what we want.

The implications of the change for OSMF are low, since they run all their servers in UTC. But the implications for anyone else running the software, for example a developer on a laptop in another timezone, is that they are likely to get a messed up timestamps in their db. Values that should be UTC (e.g in a dump) are stored as local times instead. Changing the column type to properly reflect that the timestamps are UTC and not local time will remove a whole lot of head-scratching.

So to summarise, we are currently using `timestamp without timezone` (aka "this timestamp should be interpreted as local time") and then jumping through hoops to make every client connect as UTC. A simple column type change to `timestamp with timezone` (aka "this timestamp is a UTC timestamp") would be the right thing to do.

**Comment by pnorman** [comment]
> I hope to persuade them of the case for changing the column type.

This matches what I have heard from 2-4 PostgreSQL consultants who I have asked about the matter. In summary, their advice is "Unless you are developing calendar software, **never** use `timestamp without timezone`"

**Comment by pnorman** [comment]
> @gravitystorm : please keep in mind that changing to "timestamp with time zone" possibly also has some impact on cgimap, it's not a rails only thing. This issue has hit us multiple times both on rails and cgimap in the past already.

This was raised in https://github.com/openstreetmap/openstreetmap-website/pull/2028#discussion_r235328787

If everything involve is using UTC, there's no cgimap issue. Where I've hit the issue has been on a dev machine where it's not UTC, and in that case, it's got a good chance of being screwed up.

**Comment by gravitystorm** [comment]
@pnorman exactly. The only reason to use it is for events that you want to happen in local time. So you could store an event "worldwide morning yoga sessions" at "2018-11-21T0900" `timestamp without timezone` meaning that it happens 38 times, or store "I want to wake up at 8am" as `timestamp without timezone` so that it happens at 8am in the winter and 8am at the summer. But for anything else, it's the wrong column type.

I really do think this is all a misinterpretation based on what people *think* the column types mean, based on the names of the types, rather than what they *actually* mean. I can understand the misinterpretation completely, it's not clearly expressed.

**Comment by tomhughes** [comment]
Surely "with timezone" has to store the timezone? The actual time might be normalised to UTC but it has to store the original timezone as well so it can return it. That's the whole point right?

Or have I completely misunderstood SQL for years... Time to do some reading I guess.

**Comment by gravitystorm** [comment]
> Surely "with timezone" has to store the timezone? The actual time might be normalised to UTC but it has to store the original timezone as well so it can return it. That's the whole point right?

No, that's not how it works! :-) There's no offset stored, the information about the original timezone is ignored. If you store somethings as ".....10:05Z+0800" into a `timestamp with timezone` then it just gets stored as "....18:05UTC". Like I say, I think the column type names are a bit misleading.

https://stackoverflow.com/a/9576170/105451 is another good article. It notes "The time zone itself is never stored. It is an input modifier used to compute the according UTC timestamp, which is stored - or and output modifier used to compute the local time to display - with appended time zone offset."

**Comment by tomhughes** [comment]
So it seems you're right which is very, err, counter-intuitive. My understanding had always been with a `with timezone` column stored an indication of the original timezone so that you could store values from different timezones and then recover the original zone.

Apparently that is not true for Postgres at least, and it always returns values in the servers default timezone.

My only remaining concern is that timestamp constants are always treated as `without timezone` unless explicitly cast which might cause an issue if somebody had a non-UTC default timezone?

**Comment by tomhughes** [comment]
I think this is also a point where different databases are not consistent - my reading is that Oracle for example does do what I expected but also has an extra `with local time zone` version that behaves more like Postgres does (except it normalises to the servers's time zone not necessarily UTC).

**Comment by gravitystorm** [comment]
> My only remaining concern is that timestamp constants are always treated as `without timezone` unless explicitly cast which might cause an issue if somebody had a non-UTC default timezone?

Do we use timestamp constants anywhere? I'm guessing not in our rails code, but perhaps in other apps? If so, they could be re-written to be a `timestampz` e.g. `select ('2017-01-01 00:00:00' at time zone 'utc')::timestamptz`

**Comment by tomhughes** [comment]
I have no idea whether we do (or whether rails might behind the scenes) it's just something to keep in mind.

**Comment by gravitystorm** [comment]
> I have no idea whether we do (or whether rails might behind the scenes) it's just something to keep in mind.

Noted. My remaining query is on the time taken to make the column type changes in production, i.e. will we need any downtime to run the migration, or given it should (hopefully) be just a column metadata change, is it quick? Is there an easy way to find this out before we do it for real?

**Comment by tomhughes** [comment]
That's a good question, especially in 9.5, so we need to find a way to check it.

We'd likely need a brief outage anyway to avoid lock contention issues on some of the tables at least but we'd still need to know it would be quick excluding that.

**Comment by mmd-osm** [comment]
> Where I've hit the issue has been on a dev machine where it's not UTC, and in that case, it's got a good chance of being screwed up.

Right, cgimap indeed depends on UTC-0, and there will be some screw up otherwise: https://github.com/zerebubuth/openstreetmap-cgimap/issues/150.

**Comment by tomhughes** [comment]
I think that entirely depends what timezone you want cgimap to return those timestamps in... If you want them in UTC then you're good, if you want them in local time then no.

**Comment by mmd-osm** [comment]
The "Z" (zulu) character in the timestamp string implies that all timestamps should be returned in UTC-0, no matter what time zone the server runs in. 

I also remember seeing some issues when populating the database via osmosis from a UTC+3 terminal (see this year's GSoC project: https://github.com/osmlab/yalcha/issues/11). This ended up creating wrong timestamps all over the place. Not sure, if this behaviour would be any different with osmosis and `timestamp with timezone`.

**Comment by gravitystorm** [comment]
> If you want them in UTC then you're good,

Yes, that code looks to me like it needs no changes. It's already forcing the timestamp to be considered as UTC by ending in Z with no offset shown. That will work currently on databases configured as UTC, and after the column change, on all servers regardless of their timezone.

(In fact, changing the  will fix a potential source of bugs that developers might encounter. If today at noon in California you insert a node into your local development database using raw sql (INSERT INTO current_nodes VALUES .... ( ... NOW()... ) for some development work stuff, then the timestamp would contain "2018-11-21T12:00" (i.e. local time) and cgimap would show that as 2018-11-21T12:00Z" (UTC) which is a different time entirely. But when we change the column type, the NOW() would instead store the UTC timestamp in the timestamp field, and the potential bug disappears, with no changes needed to cgimap.)

**Comment by gravitystorm** [comment]
> The "Z" (zulu) character in the timestamp string implies that all timestamps should be returned in UTC-0, no matter what time zone the server runs in.

That's not quite what it's doing. It's taking the timestamp from the database, formatting it appropriately and then sticking an actual letter Z at the end of the string (like the T, it's double quoted - it's not a formatting character like YYYY so postgres is just adding that character to the string). So there's no timezone conversion happening in the to_char method. It's taking the `timestamp without timezone` and making it look like a UTC timestamp. Which of course works fine if the server is in UTC but leads to other problems in other situations.

> This ended up creating wrong timestamps all over the place.

Yes, this is the same problem. Osmosis will have read the UTC timestamp from the XML file, handled it internally as UTC all through the pipelines, and then when writing it to the database the pg connection library will have converted that time to local time when storing it as `timestamp without timezone`. Then cgimap / rails port / whatever will have used that timestamp, assumed it was UTC and there's the problem.

**Comment by gravitystorm** [comment]
I found this in the [postgresql 9.2 release notes](https://www.postgresql.org/docs/9.2/release-9-2.html), which sounds hopeful that the table changes might be done without extended downtimes:

> Reduce need to rebuild tables and indexes for certain ALTER TABLE ... ALTER COLUMN TYPE operations (Noah Misch)
>
> Increasing the length limit for a varchar or varbit column, or removing the limit altogether, no longer requires a table rewrite. Similarly, increasing the allowable precision of a numeric column, or changing a column from constrained numeric to unconstrained numeric, no longer requires a table rewrite. Table rewrites are also avoided in similar cases involving the interval, timestamp, and timestamptz types.

I think we'll still need a short downtime to get the locks, but hopefully will be fast apart from that.

Is there a replica that this could be tested on? Or should we just start with the smaller tables and work our way up? :smiley:

**Comment by mmd-osm** [comment]
I tried `alter table current_nodes alter column timestamp type timestamptz;` on a system running on UTC+1 (Eu/Berlin). Timestamps used to be correct before, and are now off by 1 or 2 hours, depending on summer/winter time. So running this statement on a non UTC+0 system seems like a bad idea.
I'm pretty sure there are some alternatives available to get this right?

```

     id     |  latitude  | longitude  | changeset_id | visible |         timestamp          |    tile    | version 
------------+------------+------------+--------------+---------+----------------------------+------------+---------
 5002730175 |  423712670 | -834339064 |         1061 | t       | 2018-06-08 20:38:42.691478 | 1701877902 |       1


http://localhost:3000/api/0.6/node/5002730175

node id="5002730175" changeset="1061" timestamp="2018-06-08T20:38:42Z

Change timestamp file column type:

alter table current_nodes alter column timestamp type timestamptz;


     id     |  latitude  | longitude  | changeset_id | visible |           timestamp           |    tile    | version 
------------+------------+------------+--------------+---------+-------------------------------+------------+---------
 5002730175 |  423712670 | -834339064 |         1061 | t       | 2018-06-08 20:38:42.691478+02 | 1701877902 |       1

http://localhost:3000/api/0.6/node/5002730175

node id="5002730175" changeset="1061" timestamp="2018-06-08T18:38:42Z"
```

**Comment by pnorman** [comment]
> Noted. My remaining query is on the time taken to make the column type changes in production, i.e. will we need any downtime to run the migration, or given it should (hopefully) be just a column metadata change, is it quick? Is there an easy way to find this out before we do it for real?

The values need to be converted from one type to another. It's a trivial conversion, but it still needs to be done. Because the `nodes` table isn't partitioned that means the parallelism is limited.

It's probably possible to do this without an outage but not easy.

**Comment by mmd-osm** [comment]
@pnorman, @gravitystorm : what would be your take on how to address the breakage of cgimap, osmosis, and other tools due to this change on a non-UTC system?

**Comment by tomhughes** [comment]
@pnorman really, even if the system is running on UTC so that nothing changes?

If the columns need to be rewritten then I think this will pretty much be a non-starter as we would need a downtime measured in hours if not days.

**Comment by pnorman** [comment]
> @pnorman, @gravitystorm : what would be your take on how to address the breakage of cgimap, osmosis, and other tools due to this change on a non-UTC system?

My take is that you're already broken if you're running those on a non-UTC system.

The best transition plan is probably to force the timezone for the users those clients connect from, or to set appropriate libpq env vars to force the timezone.

**Comment by pnorman** [comment]
> @pnorman really, even if the system is running on UTC so that nothing changes?
> 
> If the columns need to be rewritten then I think this will pretty much be a non-starter as we would need a downtime measured in hours if not days.

This is where I'd defer to a consultant as it's a reasonably well defined problem, and sufficiently advanced that it's outside my experience.

**Comment by mmd-osm** [comment]
> The best transition plan is probably to force the timezone for the users those clients connect from, or to set appropriate libpq env vars to force the timezone.

I see this more as a hack than a real solution. in the long run, we would still require various code changes across different applications to make it work as seamlessly as today.

**Comment by mmd-osm** [comment]
> really, even if the system is running on UTC so that nothing changes?

There are a few pointers in the documentation and source code:

According to: https://doxygen.postgresql.org/tablecmds_8c_source.html#l09176 (points to relevant Postgres source code):

```
  * When the data type of a column is changed, a rewrite might not be required
  * if the new type is sufficiently identical to the old one, and the USING
  * clause isn't trying to insert some other value.  It's safe to skip the
  * rewrite if the old type is binary coercible to the new type, or if the
  * new type is an unconstrained domain over the old type.  In the case of a
  * constrained domain, we could get by with scanning the table and checking
  * the constraint rather than actually rewriting it, but we don't currently
  * try to do that.
  */
```

According to https://www.postgresql.org/docs/9.5/sql-altertable.html

_Adding a column with a DEFAULT clause or changing the type of an existing column will require the entire table and its indexes to be rewritten. As an exception when changing the type of an existing column, if the USING clause does not change the column contents and the old type is either binary coercible to the new type or an unconstrained domain over the new type, a table rewrite is not needed; but any indexes on the affected columns must still be rebuilt._

My conclusion would be (TBC):

* In a pure UTC-0 set up (no summer/winter time), the value should not change, timestamp & timestamptz are not constrained, hence no rewrite should be necessary. An index rebuild *is* necessary, though.
* In a non-UTC set up, a rewrite is necessary, as the value changes (`timestamp` has to be converted to the local timezone). Index rebuild is also necessary.


<details>
<summary>Why is the timestamp conversion not binary-coercible?</summary>

```
openstreetmap=# select * from pg_cast where castsource='timestamp'::regtype and casttarget='timestamptz'::regtype;
 castsource | casttarget | castfunc | castcontext | castmethod 
------------+------------+----------+-------------+------------
       1114 |       1184 |     2028 | i           | f
(1 row)
```


According to https://www.postgresql.org/docs/9.5/catalog-pg-cast.html

Field castmethod: Indicates how the cast is performed. 
- f means that the function specified in the castfunc field is used. 
- i means that the input/output functions are used.
- b means that the types are binary-coercible, thus no conversion is required.
</details>

**Comment by pnorman** [comment]
> I see this more as a hack than a real solution. in the long run, we would still require various code changes across different applications to make it work as seamless as today.

This change is to avoid the hack of forcing every client to specify the timezone. Right now you have to do that, in the future you won't if we make this change.

**Comment by mmd-osm** [comment]
This is not how it works today. I’m running rails, cgimap and osmosis on utc+1/2 and don’t have to mess around with time zones. It works out of the box as is.

**Comment by gravitystorm** [comment]
>  it works out of the box.

Not quite - it *appears* to work out of the box. This is because the clients you list all assume that the data stored in the 'timestamp' columns is UTC, when it is actually 'local time'. Since the assumption is symmetric (i.e. the same mistake on writing and reading), you don't see the problem. That, however, is not sufficient to state that the problem doesn't exist. :-)

> @pnorman, @gravitystorm : what would be your take on how to address the breakage of cgimap, osmosis, and other tools due to this change on a non-UTC system?

Sure, we need to fix any tools that already make broken assumptions. I'm not going to suggest we make changes that break anything, this whole topic is about fixing things properly. But this change will make the timestamp storage correct, and allow people to make future clients (or even trivial things, like calling now() from an SQL statement) work reliably without having to deal with this headache over and over for ever more.

> If the columns need to be rewritten then I think this will pretty much be a non-starter as we would need a downtime measured in hours if not days.

We can't just say that we'll never change anything in our tables ever again! If this does involve large table changes, then we can work out what the zero-downtime migration path would be, and do that. We could practise multi-stage migrations on the smaller tables where there is less impact.

**Comment by gravitystorm** [comment]
> Contrary to some earlier statements, **cgimap shows incorrect timestamps now**, in case the server isn't running on UTC+0.

OK, I mentioned to @tomhughes last week that I feared this might be the case and that it needs more investigation. From my initial investigations it appears that there isn't a version of `to_char` that takes `timestampz`, so it is most likely converting `timestampz` to `timestamp` and then doing the character formatting.

> I think we would have to add `at time zone 'UTC'` to all statements for a correct timestamp then. This change however needs to be exactly in sync with changing the column type to "timestamp with time zone" to avoid incorrect timestamps on a non-UTC system.

If there's a way to avoid the synchronous changes, that would be worth looking into. For example, cgimap could try column type detection, or we can find a different time formatting function, or so on. Basically we should try to come up with a solution that allows cgimap to work with both timestamp and timestampz columns interchangeably.
